src/twisted/application/runner/_pidfile.py

```diff
@@ -0,0 +1,202 @@
+        @type filePath: L{IFilePath}
+        """
+        self.filePath = IFilePath(filePath)
+
+
+    def read(self):
+        """
+        Read the process ID stored in this PID file.
+
+        @return: The contained process ID, or C{None} if the PID file does not
+            exist.
+        @rtype: L{int}
+
+        @raise EnvironmentError: If this PID file cannot be read.
+        @raise ValueError: If this PID file's content is invalid.
+        """
+        pidString = b""
+        try:
+            for pidString in self.filePath.open():
+                break
+        except OSError as e:
+            if e.errno == errno.ENOENT:  # No such file
+                return None
+            raise
+
+        return int(pidString)
+
+
+    def write(self, pid=None):
+        """
+        Store a PID in this PID file.
+
+        @param pid: A PID to store.  If C{None}, store the currently running
+            process ID.
+        @type pid: L{int}
+
+        @raise EnvironmentError: If this PID file cannot be written.
+        """
+        if pid is None:
+            pid = getpid()
+        else:
+            pid = int(pid)
+
+        self.filePath.setContent(self.format(pid=pid))
+
+
+    def remove(self):
+        """
+        Remove this PID file.
+
+        @raise EnvironmentError: If this PID file cannot be removed.
+        """
+        self.filePath.remove()
+
+
+    def isRunning(self):
+        """
+        Determine whether there is a running process corresponding to the PID
+        in this PID file.
+
+        @return: True if this PID file contains a PID and a process with that
+        PID is currently running; false otherwise.
+        @rtype: L{bool}
+
+        @raise EnvironmentError: If this PID file cannot be read.
+        @raise ValueError: If this PID file's content is invalid.
+        """
+        pid = self.read()
+        if pid is None:
+            return False
+
+        try:
+            kill(pid, 0)
+        except OSError as e:
+            if e.errno == errno.ESRCH:  # No such process
+                self._log.info("Removing stale PID file: {log_source}")
+                self.remove()
+                return False
+            elif e.errno == errno.EPERM:  # Not permitted to kill
+                return True
+            else:
+                raise
+        else:
+            return True
+
+
+    def __enter__(self):
+        """
+        Enter a context using this PIDFile.
+
+        Writes the PID file with the PID of the running process.
+
+        @raise AlreadyRunningError: A process corresponding to the PID in this
+            PID file is already running.
+        """
+        if self.isRunning():
+            raise AlreadyRunningError()
+        self.write()
+        return self
+
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """
+        Exit a context using this PIDFile.
+
+        Removes the PID file.
+        """
+        self.remove()
+
+
+
+class NonePIDFile(PIDFile):
```

Comment:
IPIDFile added in d36c4695c6fdd246087afad3351f9577c8ae670f, inheritance removed.